ipaserver/install/plugins/adtrust.py

```diff
@@ -716,3 +718,96 @@ def execute(self, **options):
+        result = ipautil.run([paths.KLIST, "-etK", "-k", keytab],
+                             capture_output=True, raiseonerr=False,
+                             nolog_output=True)
+        if result.returncode != 0:
+            return None
+
+        keys_to_sync = []
+        for l in result.output.splitlines():
+            if (host_princ in l and any(e in l for e in self.valid_etypes)):
+
+                els = l.split()
+                els[4] = els[4].strip('()')
+                els[5] = els[5].strip('()')
+                keys_to_sync.append(KeyEntry._make(els))
+
+        return keys_to_sync
+
+    def copy_key(self, keytab, keyentry):
+        # keyentry.key is a hex value of the actual key
+        # prefixed with 0x, as produced by klist -K -k.
+        # However, ktutil accepts hex value without 0x, so
+        # we should strip first two characters.
+        stdin = dedent("""\
+        rkt {keytab}
+        addent -key -p {principal} -k {kvno} -e {etype}
+        {key}
+        wkt {keytab}
+        """).format(keytab=keytab, principal=keyentry.principal,
+                    kvno=keyentry.kvno, etype=keyentry.etype,
+                    key=keyentry.key[2:])
+
+        result = ipautil.run([paths.KTUTIL], stdin=stdin, raiseonerr=False,
+                             umask=0o077, nolog_output=True)
+
+        if result.returncode != 0:
+            logger.warning('Unable to update %s with new keys', keytab)
+
+    def execute(self, **options):
+        # First, see if trusts are enabled on the server
+        if not self.api.Command.adtrust_is_enabled()['result']:
+            logger.debug('AD Trusts are not enabled on this server')
+            return False, []
+
+        # Extract keys from the host and samba keytabs
+        hostkeys = self.extract_key_refs(paths.KRB5_KEYTAB)
+        cifskeys = self.extract_key_refs(paths.SAMBA_KEYTAB)
+        if any([hostkeys is None, cifskeys is None]):
+            logger.warning('Either %s or %s are missing or unreadable',
+                           paths.KRB5_KEYTAB, paths.SAMBA_KEYTAB)
+            return False, []
+
+        # If there are missing host keys in the samba keytab, copy them over
+        # Also copy those keys that differ in the content and/or KVNO
+        for hostkey in hostkeys:
+            copied = False
+            uptodate = False
+            for cifskey in cifskeys:
+                if all([cifskey.principal == hostkey.principal,
+                        cifskey.etype == hostkey.etype]):
+                    if any([cifskey.key != hostkey.key,
+                            cifskey.kvno != hostkey.kvno]):
+                        self.copy_key(paths.SAMBA_KEYTAB, hostkey)
+                        copied = True
+                        break
+                    else:
+                        uptodate = True
+            if not (copied or uptodate):
+                self.copy_key(paths.SAMBA_KEYTAB, hostkey)
+
+        return False, []
```

Comment:
Does this mean if someone gets a new host keytab they will need to run this upgrade code in order to update the cifs principal? If someone does this would there be reasonably understandable error messages or just access denied?

Reply:
They can re-run ipa-adtrust-install which will copy the keytab over. I think the error message in Samba logs will show the problem  as `GSS VerifyMic failed: ...`:
(from `samba/auth/kerberos/gssapi_helper.c:NTSTATUS gssapi_check_packet()`)
```
        if (GSS_ERROR(maj_stat)) {
                char *error_string = gssapi_error_string(NULL,
                                                         maj_stat,
                                                         min_stat,
                                                         mech);
                DEBUG(1, ("GSS VerifyMic failed: %s\n", error_string));
                talloc_free(error_string);

                return NT_STATUS_ACCESS_DENIED;
        }
```

Reply:
Ok. This may lead us to verifying the two keytabs are consistent. If not we can always add something later.

Reply:
The verification is running every time we upgrade data. So any package update, for example. If not, re-running `ipa-adtrust-install` will do this for you too. Finally, making an update file `90-update-host-cifs-keytabs.update` with
```
plugin: update_host_cifs_keytabs
```
and running it with `ipa-ldap-updater ./90-update-host-cifs-keytabs.update` would do the same action.